lib/delta-table/src/Database/Table/SQL/Expr.hs

```diff
@@ -0,0 +1,211 @@
+false = Zero FalseE
+
+-- | SQL @NOT@, negation.
+not_ :: Expr Bool -> Expr Bool
+not_ = One Not
+
+-- | SQL @AND@, conjunction.
+(&&.) :: Expr Bool -> Expr Bool -> Expr Bool
+(&&.) = Two And
+
+-- | Combine multiple 'Expr' with '(&&.)'.
+and_ :: [Expr Bool] -> Expr Bool
+and_ [] = true
+and_ (x:xs) = foldr (&&.) x xs
+
+-- | SQL @OR@, disjunction.
+(||.) :: Expr Bool -> Expr Bool -> Expr Bool
+(||.) = Two Or
+
+-- | Combine multiple 'Expr' with '(||.)'.
+or_ :: [Expr Bool] -> Expr Bool
+or_ [] = false
+or_ (x:xs) = foldr (||.) x xs
+
+-- | Comparing a column against a given value.
+columnOpValue
+    :: forall n a c. (IsColumnName n, Sqlite.ToField a)
+    => Op2 Identifier Identifier c -> Col n a -> a -> Expr c
+columnOpValue op2 col a =
+    Two op2
+        (Zero . ColumnName $ getColumnName col)
+        (Value $ Sqlite.toField a)
+
+-- | Test whether the value in a given column equals a given value.
+(==.)
+    :: forall n a. (IsColumnName n, Sqlite.ToField a)
+    => Col n a -> a -> Expr Bool
+(==.) = columnOpValue Eq
+
+(/=.)
+    :: forall n a. (IsColumnName n, Sqlite.ToField a)
+    => Col n a -> a -> Expr Bool
+(/=.) col a = not_ (col ==. a)
+
+(<.)
+    :: forall n a. (IsColumnName n, Sqlite.ToField a, Ord a)
+    => Col n a -> a -> Expr Bool
+(<.) = columnOpValue Lt
+
+(<=.)
+    :: forall n a. (IsColumnName n, Sqlite.ToField a, Ord a)
+    => Col n a -> a -> Expr Bool
+(<=.) = columnOpValue Leq
+
+(>.)
+    :: forall n a. (IsColumnName n, Sqlite.ToField a, Ord a)
+    => Col n a -> a -> Expr Bool
+(>.) = columnOpValue Gt
+
+(>=.)
+    :: forall n a. (IsColumnName n, Sqlite.ToField a, Ord a)
+    => Col n a -> a -> Expr Bool
+(>=.) = columnOpValue Geq
+
+{-------------------------------------------------------------------------------
+    Expressions
+    Rendering
+-------------------------------------------------------------------------------}
+-- | Render an expression as SQL source code.
+renderExpr :: Expr a -> Var.Lets Text
+renderExpr (Value value) =
+    Var.bindValue value >>= var
+renderExpr (Zero e) =
+    text (render0 e)
+renderExpr (One op1 ea) =
+    text (renderOp1 op1)
+        <> sp <> parens (renderExpr ea)
+renderExpr (Two op2 ea eb) =
+    parens (renderExpr ea)
+        <> sp <> text (renderOp2 op2)
+        <> sp <> parens (renderExpr eb)
+
+-- | A single whitespace character.
+sp :: Var.Lets Text
+sp = text " "
+
+mkIdentifier :: Text -> Text
+mkIdentifier name = "\"" <> name <> "\""
+
+render0 :: Expr0 a -> Text
+render0 TrueE = "true"
+render0 FalseE = "false"
+render0 (ColumnName name) = mkIdentifier name
+
+renderOp1 :: Op1 a b -> Text
+renderOp1 Not = "NOT"
+
+renderOp2 :: Op2 a b c -> Text
+renderOp2 And = "AND"
+renderOp2 Or = "OR"
+renderOp2 Eq = "="
+renderOp2 Lt = "<"
+renderOp2 Leq = "<="
+renderOp2 Gt = ">"
+renderOp2 Geq = ">="
+
+-- | Render a fixed text.
+text :: Text -> Var.Lets Text
+text = pure
+
+-- | Render a variable name
+var :: Var.VarName -> Var.Lets Text
+var = pure
```

Comment:
Worth making `VarName` a `newtype` to distinguish `var` from `text`?

Reply:
🤔 Yes. Making the type `VarName` abstract is a good idea, because the only sensible way to create variable names should be `newVarName` anyway.